Intrinsic: introduce minimumnum and maximumnum for IR and SelectionDAG

wzssyqa opened the pull request:
C23 introduced new functions fminimum_num and fmaximum_num, and they follow the minimumNumber and maximumNumber of IEEE754-2019. Let's introduce new intrinsics to support them.

This patch introduces support only support for scalar values. The support of
  vector (vp, vp.reduce, vector.reduce),
  experimental.constrained
will be added in future patches.

With this patch, MIPSr6 and LoongArch can work out of box with fcanonical and fmax/fmin.

Aarch64/PowerPC64 can use the same login as MIPSr6 and LoongArch, while they have no fcanonical support yet.
I will add it in future patches.

The FMIN/FMAX of RISC-V instructions follows the minimumNumber/maximumNumber of IEEE754-2019. We can just add it in future patch.

Background
https://discourse.llvm.org/t/rfc-fix-llvm-min-f-and-llvm-max-f-intrinsics/79735
Currently we have fminnum/fmaxnum, which have different behavior on different platform for NUM vs sNaN:
   1) Fallback to fmin(3)/fmax(3): return qNaN.
   2) ARM64/ARM32+Neon: same as libc.
   3) MIPSr6/LoongArch/RISC-V: return NUM.

And the fix of fminnum/fmaxnum to follow minNUM/maxNUM of IEEE754-2008 will submit as separated patches.

wzssyqa committed:
Intrinsic: introduce minimumnum and maximumnum for IR and SelectionDAG

C23 introduced new functions fminimum_num and fmaximum_num,
and they follow the minimumNumber and maximumNumber of IEEE754-2019.
Let's introduce new intrinsics to support them.

This patch introduces support only support for scalar values.
The support of
  vector (vp, vp.reduce, vector.reduce),
  experimental.constrained
will be added in future patches.

With this patch, MIPSr6 and LoongArch can work out of box.
Other architectures will fallback to libc call. The support of them
will be added in future patches.

Background
Currently we have fminnum/fmaxnum, which have different
behavior on different platform for NUM vs sNaN:
   1) Fallback to fmin(3)/fmax(3): return qNaN.
   2) ARM64/ARM32+Neon: same as libc.
   3) MIPSr6/LoongArch/RISC-V: return NUM.

And the fix of fminnum/fmaxnum to follow minNUM/maxNUM of IEEE754-2008
will submit as separated patches.

tschuett commented:
Why is APFloat in this patch?

wzssyqa commented:
> Why is APFloat in this patch?

To make the CI happy.  I will rebase it once https://github.com/llvm/llvm-project/pull/96304 is merged.

wzssyqa commented:
@nikic @arames gently ping.

arsenm commented on llvm/docs/LangRef.rst:
```diff
@@ -16102,6 +16102,100 @@ The returned value is completely identical to the input except for the sign bit;
+Only basic intrinsics list here. The constrained version
+ones may have different behaivor on exception.
+
+Since some architectures implement minNum/maxNum with +0.0>-0.0,
+so we define internal ISD::MINNUM_IEEE and ISD::MAXNUM_IEEE.
+They will be helpful to implement minimumnum/maximumnum.
```
This description doesn't belong in the LangRef unless they are IR visible. Also the primary difference for these is the signaling nan handling matches the 2008 behavior

wzssyqa replied:
OK. Let's drop it.

I mean that we can use
```
fcanonical
fcanonical
MINNUM_IEEE
```
to implement `minimumnum` on such architectures, like ARM64/MIPSr6/LoongArch/PowerPC.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGDumper.cpp:
```diff
@@ -203,6 +203,8 @@ std::string SDNode::getOperationName(const SelectionDAG *G) const {
   case ISD::STRICT_FMINIMUM:            return "strict_fminimum";
   case ISD::FMAXIMUM:                   return "fmaximum";
   case ISD::STRICT_FMAXIMUM:            return "strict_fmaximum";
+  case ISD::FMINIMUMNUM:               return "fminimumnum";
+  case ISD::FMAXIMUMNUM:               return "fmaximumnum";
```
Align to other cases

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8469,6 +8469,93 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
   return MinMax;
 }
 
+SDValue TargetLowering::createSelectForFMINIMUMNUM_FMAXIMUMNUM(
```
This probably shouldn't be surfaced as a public can-fail helper

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8469,6 +8469,93 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+  if (SDValue SelCC = createSelectForFMINIMUMNUM_FMAXIMUMNUM(Node, DAG))
+    return SelCC;
+
+  return SDValue();
```
This should be implemented as to never fail

wzssyqa committed:
Remove MINNUM_IEEE in LangRef

wzssyqa committed:
fix code style

wzssyqa committed:
remove createSelectForFMINIMUMNUM_FMAXIMUMNUM

wzssyqa committed:
fix code style

Urgau commented:
I don't know much about LLVM but why is it introducing new intrinsic while `llvm.minimum.*` and `llvm.maximum.*` already claim to follow the "IEEE 754-2019" definition and has already support in a few backends?

> If either operand is a NaN, returns NaN. Otherwise returns the greater of the two arguments. -0.0 is considered to be less than +0.0 for this intrinsic. Note that these are the semantics specified in the draft of IEEE 754-2019.

arsenm commented:
IEEE-754 2019 has separate minimum and minimumNumber operations

wzssyqa commented:
Gently ping @arsenm

arsenm commented on llvm/docs/LangRef.rst:
```diff
@@ -16101,6 +16101,96 @@ The returned value is completely identical to the input except for the sign bit;
+LLVM Implementation:
+""""""""""""""""""""
+
+LLVM implements all ISO C flavors as listed in this table.
+Only basic intrinsics list here. The constrained version
+ones may have different behaivor on exception.
```
```suggestion
LLVM implements all ISO C flavors as listed in this table, except in the default floating-point environment exceptions are ignored. The constrained versions of the intrinsics respect the exception behavior. 

```

arsenm commented on llvm/docs/LangRef.rst:
```diff
@@ -16282,6 +16372,98 @@ of the two arguments. -0.0 is considered to be less than +0.0 for this
+
+Note that these are the semantics of minimumNumber specified in IEEE 754-2019.
+
+It has some different with '``llvm.minnum.*``':
```
```suggestion
It has some differences with '``llvm.minnum.*``':
```

arsenm commented on llvm/docs/LangRef.rst:
```diff
@@ -16282,6 +16372,98 @@ of the two arguments. -0.0 is considered to be less than +0.0 for this
+
+Note that these are the semantics of maximumNumber specified in IEEE 754-2019.
+
+It has some different with '``llvm.maxnum.*``':
```
```suggestion
It has some differences with '``llvm.maxnum.*``':
```

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8531,6 +8531,70 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+                                                         : ISD::FMAXNUM_IEEE;
+  EVT VT = Node->getValueType(0);
+
+  if (VT.isScalableVector())
+    report_fatal_error(
+        "Expanding fminimumnum/fmaximumnum for scalable vectors is undefined.");
```
It's not undefined. But I don't see how scalable vectors are relevant here. The code should work as-is, this just needs to use the same type for every operation

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8531,6 +8531,70 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+                         Node->getOperand(1), Node->getFlags());
+  }
+
+  // FIXME: we may use setCC/getSELECT to optimize it instead of fallback to
+  // libcall.
```
This is not an optimization, do not refer to it as such. No libcall should be emitted here either, this is for the inline expansion

wzssyqa replied:
libcall is not emitted here. Sorry for me that I didn't descriptor clearly.
I mean if we return `SDValue()`, the later code will emit libcall, while if we return setCC/getSELECT, no libcall will be used later.

Yes, maybe libcall is not so expensive, so that we can be happy to use it.

arsenm replied:
This is the wrong way to structure it. TLI::expand* functions should always fully, and directly, expand in terms of other generic opcodes. It should not have a failure path, expecting other code to insert a libcall on failure. There should be no failure case here. Other code should explicitly choose to use the libcall

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8531,6 +8531,70 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+
+  // FIXME: we may use setCC/getSELECT to optimize it instead of fallback to
+  // libcall.
+  return SDValue();
```
This should still have a never fail path

wzssyqa replied:
Do you mean that we should do like `expandFMINIMUM_FMAXIMUM`, so that there is no `fminimum_num(3)` is emit here?

arsenm commented on llvm/test/CodeGen/Mips/fp-maximumnum-minimumnum.ll:
```diff
@@ -0,0 +1,100 @@
+; MIPS32-NEXT:    addiu $sp, $sp, 24
+  %z = call double @llvm.minimumnum.f64(double %x, double %y)
+  ret double %z
+}
```
Missing all the vector tests

wzssyqa replied:
Do you mean something like 

```
define void @fminimumnum_v4f32(ptr %x, ptr %y) {
  %a = load <4 x float>, ptr %x
  %b = load <4 x float>, ptr %y
  %c = tail call <4 x float> @llvm.minimum.f32 (<4 x float> %a, <4 x float> %b)
  store <4 x float> %c, ptr %x
  ret void
}
```

If so, with this patch, LLVM will crash with
```
fatal error: error in backend: Do not know how to split the result of this operator!
```

As we haven't add vector support yet. To make one patch short, I split it out.

arsenm reviewed (commented):
I do figure out a version like this

```
+  SDValue MinMax;
+  SDValue LHSCompareLHS = DAG.getSetCC(DL, CCVT, LHS, LHS, ISD::SETEQ);
+  LHS = DAG.getSelect(DL, VT, LHSCompareLHS, LHS, RHS, Flags);
+  SDValue RHSCompareRHS = DAG.getSetCC(DL, CCVT, RHS, RHS, ISD::SETEQ);
+  RHS = DAG.getSelect(DL, VT, RHSCompareRHS, RHS, LHS, Flags);
+  SDValue Compare =
+      DAG.getSetCC(DL, CCVT, LHS, RHS, IsMax ? ISD::SETGT : ISD::SETLT);
+  MinMax = DAG.getSelect(DL, VT, Compare, LHS, RHS, Flags);
+  SDValue MinMaxCompareMinMax = DAG.getSetCC(DL, CCVT, MinMax, MinMax, ISD::SETEQ);
+  SDValue MinMaxQuiet = DAG.getNode(ISD::FADD, DL, VT, MinMax, DAG.getConstantFP(0.0, DL, VT), Flags);
+  MinMax = DAG.getSelect(DL, VT, MinMaxCompareMinMax, MinMax, MinMaxQuiet, Flags);
+
+  SDValue TestZero =
+        DAG.getTargetConstant(IsMax ? fcPosZero : fcNegZero, DL, MVT::i32);
+  SDValue IsZero = DAG.getSetCC(DL, CCVT, MinMax,
+                                DAG.getConstantFP(0.0, DL, VT), ISD::SETEQ);
+  SDValue LCmp = DAG.getSelect(
+        DL, VT, DAG.getNode(ISD::IS_FPCLASS, DL, CCVT, LHS, TestZero), LHS,
+        MinMax, Flags);
+  SDValue RCmp = DAG.getSelect(
+        DL, VT, DAG.getNode(ISD::IS_FPCLASS, DL, CCVT, RHS, TestZero), RHS,
+        LCmp, Flags);
+  MinMax = DAG.getSelect(DL, VT, IsZero, RCmp, MinMax, Flags);
+
+  return MinMax;
```

While it has only 1/2 performance of libcall on x86_64.
The reason is that libcall can just return if the value is comparable.
I have no idea how to use branch instructions here.  Any idea?

wzssyqa committed:
improve LangRef.rst

wzssyqa committed:
Use SetCC/SelectCC as final fallback

It seems exposing a bug of RISC-V backend:

```
define dso_local float @f(float noundef %a, float noundef %b) local_unnamed_addr #0 {
entry:
  %0 = tail call float @llvm.minimumnum.f32(float %a, float %b)
  ret float %0
}
```
```
./bin/clang --target=riscv64-linux-gnu -S yy.ll -O2 -o -
        li      a0, 1
        bnez    a0, .LBB0_2
```

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8531,6 +8531,94 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+      return DAG.getNode(IEEE2008Op, DL, VT, LHS, RHS, Flags);
+  }
+
+  SDValue MinMax;
```
Declare on first define. Also would be better to avoid mutating the same variable for every step

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8531,6 +8531,94 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+  MinMax = DAG.getSelect(DL, VT, Compare, LHS, RHS, Flags);
+
+  // If MinMax is NaN, let's quiet it with MinMax = MinMax + 0.
+  SDValue MinMaxCompareMinMax =
+      DAG.getSetCC(DL, CCVT, MinMax, MinMax, ISD::SETEQ);
+  SDValue MinMaxQuiet = DAG.getNode(ISD::FADD, DL, VT, MinMax,
+                                    DAG.getConstantFP(0.0, DL, VT), Flags);
+  MinMax =
+      DAG.getSelect(DL, VT, MinMaxCompareMinMax, MinMax, MinMaxQuiet, Flags);
```
I guess you could use getSelectCC here, but it makes little difference

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -3650,6 +3650,12 @@ bool SelectionDAGLegalize::ExpandNode(SDNode *Node) {
   }
+  case ISD::FMINIMUMNUM:
+  case ISD::FMAXIMUMNUM: {
+    if (SDValue Expanded = TLI.expandFMINIMUMNUM_FMAXIMUMNUM(Node, DAG))
```
Remove the check, it should always succeed

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8531,6 +8531,94 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+  }
+
+  // FMINIMUM/FMAXIMUM always return NaN if either operand is NaN.
+  // It has same behavior about +0.0 vs -0.0.
```
Reword this (don't know what about means here), and same as what?

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8531,6 +8531,94 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+    if (!Flags.hasNoNaNs()) {
+      // Insert canonicalizes if it's possible we need to quiet to get correct
+      // sNaN behavior.
+      if (!DAG.isKnownNeverSNaN(LHS)) {
+        LHS = DAG.getNode(ISD::FCANONICALIZE, DL, VT, LHS, Node->getFlags());
+      }
+      if (!DAG.isKnownNeverSNaN(RHS)) {
+        RHS = DAG.getNode(ISD::FCANONICALIZE, DL, VT, RHS, Node->getFlags());
+      }
```
```suggestion
      if (!DAG.isKnownNeverSNaN(LHS))
        LHS = DAG.getNode(ISD::FCANONICALIZE, DL, VT, LHS, Flags);
      if (!DAG.isKnownNeverSNaN(RHS))
        RHS = DAG.getNode(ISD::FCANONICALIZE, DL, VT, RHS, Flags);
```

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8531,6 +8531,94 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+  MinMax =
+      DAG.getSelect(DL, VT, MinMaxCompareMinMax, MinMax, MinMaxQuiet, Flags);
+
+  // Let's process +0.0 vs -0.0.
```
```suggestion
  // Fixup signed zero behavior
```

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8531,6 +8531,94 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+
+  // FMINNUM/FMAXMUM returns qNaN if either operand is sNaN, and it may return
+  // either one for +0.0 vs -0.0.
+  // FIXME: maybe we need hasNoSNaNs().
```
FIXME is out of place, we should have this mentioned in isKnownNeverSNaN already 
```suggestion
```

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8531,6 +8531,94 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+  // If one operand is NaN, let's move another value to it.
+  // So that if only one operand is NaN, we can have both two operands are
+  // non-NaN now.
+  SDValue LHSCompareLHS = DAG.getSetCC(DL, CCVT, LHS, LHS, ISD::SETEQ);
```
```suggestion
  SDValue LHSIsNaN = DAG.getSetCC(DL, CCVT, LHS, Zero, ISD::SETUNO);
```

Name nan checks isNaN. The canonical way of testing for nan is setuno with a constant zero

wzssyqa replied:
Is `SETUO` with self OK.  Comparing with float zero may need one more temp register.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8531,6 +8531,94 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+
+  SDValue MinMax;
+  // If one operand is NaN, let's move another value to it.
+  // So that if only one operand is NaN, we can have both two operands are
+  // non-NaN now.
```
Comment is confusing

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8531,6 +8531,94 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+  }
+
+  SDValue MinMax;
+  // If one operand is NaN, let's move another value to it.
```
```suggestion
  // If one operand is NaN, let's move another value to it.
```
Nothing is moving, rephrase?

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8531,6 +8531,94 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+      DAG.getSetCC(DL, CCVT, LHS, RHS, IsMax ? ISD::SETGT : ISD::SETLT);
+  MinMax = DAG.getSelect(DL, VT, Compare, LHS, RHS, Flags);
+
+  // If MinMax is NaN, let's quiet it with MinMax = MinMax + 0.
```
fadd does not guarantee quieting, you need to use canonicalize for this, like above

wzssyqa replied:
There are couple of ports that don't support `fcanonicalize`, including x86-64, riscv64, mips pre-r6, powerpc etc.
I even tred to use `isOperationLegal(ISD::FCANONICALIZE, VT)`, while it always returns `true`,  I haven't no idea why.

wzssyqa replied:
Glibc does use `add` to quiet sNaNs, It can do so due that `add` will trap if either operand is sNaN, and the kernel will quiet the result, which is required by IEE754.
For embedded system without such kernel, it will not be possible.

wzssyqa replied:
I use `fcanonicalize` with `FMAXNUM_IEEE`, due to that I find that every arch that has `FMAXNUM_IEEE`, also has `fcanonicalize`.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8531,6 +8531,94 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+  SDValue RCmp = DAG.getSelect(
+      DL, VT, DAG.getNode(ISD::IS_FPCLASS, DL, CCVT, RHS, TestZero), RHS, LCmp,
+      Flags);
+  MinMax = DAG.getSelect(DL, VT, IsZero, RCmp, MinMax, Flags);
+
+  return MinMax;
```
directly return?

wzssyqa committed:
Remove the check in LegalizeDAG.cpp: not needed

wzssyqa committed:
Some improve to expand

1. No need to emit move if we know one operand is not NaN.
2. No need to emit quiet if we know the result of MinMax is not NaN.
3. No need to fixup signed zero behavior if we know either operand
   is not zero or Options.NoSignedZerosFPMath.

wzssyqa committed:
Round RHS/LHS if IS_FPCLASS cannot process them

wzssyqa committed:
Fix testcases

Add test for Sparc, SystemZ, X86.

ARM64: we can use fmin if we add FMINNUM_IEEE.
RISCV: fmin is good for FMINIMUMNUM.
PowerPC: Need some updates of FCANONICAL.
These architectures will be fixed in the future PRs.

wzssyqa committed:
fix code style

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8599,23 +8600,33 @@ SDValue TargetLowering::expandFMINIMUMNUM_FMAXIMUMNUM(SDNode *Node,
+      RRound = DAG.getNode(ISD::FP_ROUND, DL, MVT::f32, RHS,
+                    DAG.getIntPtrConstant(0, DL, /*isTarget=*/true));
+      RCCVT = getSetCCResultType(DAG.getDataLayout(), *DAG.getContext(), MVT::f32);
+    }
```
Do not add target checks or workarounds to this code

arsenm replied:
The actual condition would more be the integer bitcast of the FP type is not legal

wzssyqa replied:
> The actual condition would more be the integer bitcast of the FP type is not legal
Yes.

> Do not add target checks or workarounds to this code
Another consideration of this condition test is that `expandIS_FPCLASS +0/-0` is too heavy for us. 
We only need to know is the sign bit, while expandIS_FPCLASS will compare with all of the bits with +0/-0.

arsenm commented on llvm/test/CodeGen/X86/fp-maximumnum-minimumnum.ll:
```diff
@@ -0,0 +1,2121 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc --mtriple=x86_64 < %s | FileCheck %s --check-prefix=X86-64
+; RUN: llc --mtriple=i386 < %s | FileCheck %s --check-prefix=I386
+; RUN: llc --mtriple=i386 -mattr=+sse2 < %s | FileCheck %s --check-prefix=SSE2
+; RUN: llc --mtriple=i386 -mattr=+avx < %s | FileCheck %s --check-prefix=AVX
```
I'd rather just leave x86_32 untested until #100967 has a separate fix

wzssyqa committed:
Use condition integer bitcast of the FP type is not legal

wzssyqa committed:
Remove i386 tests

wzssyqa committed:
fix sparc test

wzssyqa committed:
fix code style

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8531,6 +8531,108 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+  if (Options.NoSignedZerosFPMath || Flags.hasNoSignedZeros() ||
+      DAG.isKnownNeverZeroFloat(LHS) || DAG.isKnownNeverZeroFloat(RHS)) {
+    return MinMax;
+  } else {
```
No else after return

wzssyqa replied:
Sorry, I  cannot understand it.
Do you means that the `else` after `return MinMax` makes none sense?
If so, in fact, it is useful.

If we can be sure that either operand is non-zero or we don't care about signed-zero, we can just return with process +0 vs -0 problem.

tschuett replied:
It is forbidden:
https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

wzssyqa replied:
Oh. Thanks I see.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8531,6 +8531,110 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+  // If MinMax is NaN, let's quiet it.
+  if (!Flags.hasNoNaNs() && !DAG.isKnownNeverNaN(LHS) &&
+      !DAG.isKnownNeverNaN(RHS)) {
+    // Only use FADD to quiet it if it is NaN, because -0.0+0.0->+0.0.
+    SDValue MinMaxQuiet = DAG.getNode(ISD::FADD, DL, VT, MinMax, MinMax, Flags);
```
ping

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8531,6 +8531,107 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+  SDValue LRound = LHS;
+  SDValue RRound = RHS;
+  EVT RCCVT = CCVT;
+  // expandIS_FPCLASS is buggy for GPR32+FPR64. Let's round them to single for
+  // this case.
+  if (!isOperationLegal(ISD::BITCAST, VT.changeTypeToInteger())) {
+    LRound = DAG.getNode(ISD::FP_ROUND, DL, MVT::f32, LHS,
+                         DAG.getIntPtrConstant(0, DL, /*isTarget=*/true));
+    RRound = DAG.getNode(ISD::FP_ROUND, DL, MVT::f32, RHS,
+                         DAG.getIntPtrConstant(0, DL, /*isTarget=*/true));
+    RCCVT =
+        getSetCCResultType(DAG.getDataLayout(), *DAG.getContext(), MVT::f32);
+  }
```
Please remove all workarounds from this patch. Any workaround should be done separately. You can drop x86_32 tests and any other targets where this is broken from this patch

tschuett commented:
Given the size and churn on this patch, I have a slight preference to split it into two patches : IR and backend.

wzssyqa committed:
remove dirty hack about fadd/fround

wzssyqa committed:
Remove test for Sparc/SystemZ/X86/MIPS pre-R6

wzssyqa commented:
> Given the size and churn on this patch, I have a slight preference to split it into two patches : IR and backend.

I guess that it need a minimum use case to compose a full patch.

arsenm commented:
> > Given the size and churn on this patch, I have a slight preference to split it into two patches : IR and backend.
> 
> I guess that it need a minimum use case to compose a full patch.

It's perfectly fine to only define the intrinsic in one patch, and implement codegen in another

arsenm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -8531,6 +8531,94 @@ SDValue TargetLowering::expandFMINIMUM_FMAXIMUM(SDNode *N,
+        DAG.getSelectCC(DL, MinMax, MinMax, MinMaxQuiet, MinMax, ISD::SETUO);
+  }
+
+  // Fixup signed zero behavior.
```
It would be good if we could find an implementation that doesn't rely on IS_FPCLASS. Can we do something with signbit? I think it would avoid needing to deal with the x86 issue

wzssyqa replied:
> It would be good if we could find an implementation that doesn't rely on IS_FPCLASS. Can we do something with signbit? I think it would avoid needing to deal with the x86 issue

Yes. X86 has `FGETSIGN` as Custom. Let's have a try with it.

wzssyqa replied:
> > It would be good if we could find an implementation that doesn't rely on IS_FPCLASS. Can we do something with signbit? I think it would avoid needing to deal with the x86 issue
> 
> Yes. X86 has `FGETSIGN` as Custom. Let's have a try with it.

But it seems that no other architecture defines it.

wzssyqa replied:
Any idea about how to quiet a sNaN on x86 etc?
I see this paragraph in IEEE754-2008/2019:
```
Implementations shall provide the following formatOf general-computational operations, for destinations of
all supported arithmetic formats, and, for each destination format, for operands of all supported arithmetic
formats with the same radix as the destination format. These operations shall not propagate non-canonical
results.
```
```
Under default exception handling, any operation signaling an invalid operation exception and for which a
floating-point result is to be delivered shall deliver a quiet NaN.
```

Maybe, `fadd` is OK to quiet a sNaN.

arsenm replied:
You must use canonicalize. Non strict llvm operations do not guarantee quieting

arsenm reviewed (approved):
LGTM. I'm ambivalent on splitting the IR and codegen pieces into separate patches
